File: Dev/unitTest_Engine_cpp_gtest/ObjectSystem/LayersFlag.cpp
#include <gtest/gtest.h>
#include <ace.h>
#include "../EngineTest.h"

using namespace std;
using namespace ace;

class ObjectSystem_LayersFlag : public EngineTest
{
	std::shared_ptr<Scene>	scene;
	std::shared_ptr<Layer2D> layer;

	class MyObject : public TextureObject2D
	{
	public:
		MyObject(float x, float y)
		{
			auto texture = ace::Engine::GetGraphics()->CreateTexture2D(ToAString("Data/Texture/Cloud1.png").c_str());
			SetTexture(texture);
			SetPosition(Vector2DF(x, y));
			SetCenterPosition(Vector2DF(128, 128));
		}

	protected:
		void OnUpdate()
		{
			SetAngle(GetAngle() + 2);
		}
	};

public:
	ObjectSystem_LayersFlag(bool isOpenGLMode)
		: EngineTest(ace::ToAString("LayersFlag"), isOpenGLMode, 30)
	{
	}

protected:
	void OnStart()
	{
		scene = make_shared<Scene>();
		layer = make_shared<Layer2D>();
		auto notDrawnLayer = make_shared<Layer2D>();
		auto notUpdatedLauer = make_shared<Layer2D>();
		auto object1 = make_shared<ObjectSystem_LayersFlag::MyObject>(100, 128);
		auto object2 = make_shared<ObjectSystem_LayersFlag::MyObject>(300, 128);
		auto object3 = make_shared<ObjectSystem_LayersFlag::MyObject>(500, 128);

		notDrawnLayer->SetIsDrawn(false);
		notUpdatedLauer->SetIsUpdated(false);

		layer->AddObject(object1);
		notDrawnLayer->AddObject(object2);
		notUpdatedLauer->AddObject(object3);
		scene->AddLayer(layer);
		scene->AddLayer(notDrawnLayer);
		scene->AddLayer(notUpdatedLauer);
		ace::Engine::ChangeScene(scene);
	}

	void OnUpdating()
	{
		// レイヤー削除テスト
		if (GetTime() == GetExitTime())
		{
			scene->RemoveLayer(layer);
		}
	}
};

ENGINE_TEST(ObjectSystem, LayersFlag)

File: Dev/unitTest_Engine_cs/ObjectSystem2D/LayersFlag.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using asd;

namespace unitTest_Engine_cs.ObjectSystem2D
{
	class LayersFlag : EngineTest
	{
		Scene scene;
		Layer2D layer;

		class MyObject : TextureObject2D
		{
			public MyObject( float x, float y )
			{
				Position = new Vector2DF( x, y );
			}

			protected override void OnStart()
			{
				Texture = Engine.Graphics.CreateTexture2D( CloudTexturePath );
				CenterPosition = new Vector2DF( Texture.Size.X / 2, Texture.Size.Y / 2 );
			}

			protected override void OnUpdate()
			{
				Angle += 2;
			}
		}

		public LayersFlag()
			: base(30)
		{
		}

		protected override void OnStart()
		{
			scene = new Scene();
			layer = new Layer2D();
			var notDrawnLayer = new Layer2D();
			var notUpdatedLayer = new Layer2D();
			var object1 = new MyObject( 100, 128 );
			var object2 = new MyObject( 300, 128 );
			var object3 = new MyObject( 500, 128 );

			notDrawnLayer.IsDrawn = false;
			notUpdatedLayer.IsUpdated = false;

			layer.AddObject( object1 );
			notDrawnLayer.AddObject( object2 );
			notUpdatedLayer.AddObject( object3 );
			scene.AddLayer( layer );
			scene.AddLayer( notDrawnLayer );
			scene.AddLayer( notUpdatedLayer );

			Engine.ChangeScene( scene );
		}

		protected override void OnUpdating()
		{
			// レイヤー削除テスト
			if (Time == ExitTime)
			{
				scene.RemoveLayer(layer);
			}
		}
	}
}
